src/main/java/org/ossgang/commons/observable/mappers/buffer/BufferMapper.java

```diff
@@ -0,0 +1,78 @@
+package org.ossgang.commons.observable.mappers.buffer;
+
+import static org.ossgang.commons.observable.SubscriptionOptions.ON_CHANGE;
+
+import java.util.List;
+import java.util.function.Function;
+
+import org.ossgang.commons.observable.Observable;
+import org.ossgang.commons.observable.Observer;
+import org.ossgang.commons.property.Properties;
+import org.ossgang.commons.property.Property;
+
+/**
+ * Represents a buffer that can be used as a mapping function for {@link Observable}s.
+ * <p>
+ * This buffer has properties himself, by which its behaviour can be controlled during runtime:
+ * <ul>
+ * <li>{@link #size()}: Defines the size of the buffer</li>
+ * <li>{@link #minEmitSize()}: Defines the size at which the downstream observable starts emitting. The default is
+ * {@value #DEFAULT_MIN_EMIT_SIZE}.
+ * </ul>
+ * Additionally, the buffer exposes an {@link Observer}, {@link #clearTrigger()} that can be attached to another
+ * observable, to trigger clearing of this buffer.
+ */
+public class BufferMapper<T> implements Function<T, List<T>> {
+
+    private static final int DEFAULT_MIN_EMIT_SIZE = 2;
+    private final ConcurrentCircularBuffer<T> buffer;
+    private final Property<Integer> size;
+    private final Property<Integer> minEmitSize;
+    private final Observer<?> clearTrigger;
+
+    BufferMapper(int initialBufferSize) {
+        this.buffer = new ConcurrentCircularBuffer<>();
+        this.buffer.setLength(initialBufferSize);
+        size = Properties.property(initialBufferSize);
+        minEmitSize = Properties.property(Integer.min(initialBufferSize, DEFAULT_MIN_EMIT_SIZE));
+        size.subscribe(buffer::setLength, ON_CHANGE);
+        this.clearTrigger = (v) -> buffer.clear();
+    }
+
+    @Override
+    public List<T> apply(T t) {
+        buffer.add(t);
+        List<T> buffered = buffer.toList();
+        if (buffered.size() > minEmitSize.get()) {
+            return buffered;
+        }
+        return null;
```

Comment:
This will lead to an exception when the Optional::of is called in map(). With the current implementation, a mapper must not return null, it must transform each element into exactly one other element.

Perhaps we should expose the underlying function of "derive(this, ...)" in the interface as well, which maps to Optional and therefore allows filtering and mapping in one.
Or should we allow to return null in mappers, and have filter() as a special case of map()? 
In this case we should get rid of the explicit Optional<> in the DerivedObservableValue.

Reply:
Agree. I think a map and filter operator could be handy.. but this we could solve at the ObservableValue level by providing a default method `ObservableValue#mapAndFilter(Function<I, Optional<O>)` (unfortunately with a different name).

In the case of the buffer, I would actually go for another default method that does the buffering so that the user doesn't have to know about yet another factory method and it is aligned with the reactive streams apis in other frameworks. Something like: `ObservableValue#buffer(...)`. In this case you have the problem with the settings (emit size, clear) but can be solved by making the BufferMapper extending ObservableValue and then the `#buffer` method will return a, e.g., BufferedObservableValue. So if you're interested you have access to your properties for customizing the buffer, otherwise it behaves just like a normal observable :smile: 
Of course then you have a default method on an interface that returns a direct implementation of it...... not so clean conceptually probably
What do you think guys?
I would not get rid of the null in map, I think it's a nice idea to have the guarantee that a value is always produced..

Reply:
Ok, for the moment I went for exposing the derive method on observable and observableValue (only the simple one for the moment). Have a look how you like it.

Reply:
:+1:

Reply:
Looks good. For some point in the future, I would have another idea - instead of exposing map/filter/derive, just expose derive (or "transform"?) in the interface, then have a static repository of e.g. Transformations.map(), Transformations.filter(), Transformations.buffer() etc.

For the moment 3 methods are fine, I'm just wondering what might come up more in the future ...

Reply:
Ah, sorry, I overlooked Andreas comment...
I did not call it #mapAndFilter, but just exposed the derive. Good enough for the moment?

concerning the #buffer() method on the observable ... indeed it is the other option ... probably you tw can think a bit about the pros and cons for both over a coffee? As it is now, it is a bit more detangled and it keeps the observable interfaces small ... on the other hand of course you are right ... yet another factory method...
... as we might have more, probably one could have one class with factory methods at least (e.g. something like `Operands`. (not Buffers, Filters, Collectors.... ...)
...

Reply:
Another option is, to  - in principle - only have one method on the observable ... probably call it `then()`... (instead of derive)....

Then something like this would be possible:
```
import static Operands.*;

observable.then(map(v -> v)).then(buffer(10)).then(...)...

```